bigtable/client/metadata_update_policy.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/metadata_update_policy.h"
+#include "bigtable/client/internal/make_unique.h"
+
+#include <sstream>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+MetadataParamTypes const MetadataParamTypes::PARENT("parent");
+MetadataParamTypes const MetadataParamTypes::NAME("name");
+MetadataParamTypes const MetadataParamTypes::TABLE_NAME("table_name");
+
+MetadataUpdatePolicy::MetadataUpdatePolicy(
+    std::string resource_name, MetadataParamTypes metadata_param_type) {
+  google_cloud_resource_prefix_ =
+      std::make_pair("google-cloud-resource-prefix", resource_name);
+
+  std::string value;
+  value = metadata_param_type.getType();
+  value += "=";
+  value += resource_name;
+  x_google_request_params_ = std::make_pair("x-goog-request-params", value);
+}
+
+MetadataUpdatePolicy::MetadataUpdatePolicy(
+    std::string resource_name, MetadataParamTypes metadata_param_type,
+    std::string table_id) {
+  google_cloud_resource_prefix_ =
+      std::make_pair("google-cloud-resource-prefix", resource_name);
+
+  std::string value;
+  value = metadata_param_type.getType();
+  value += "=";
+  value += resource_name;
+  value += "/tables/" + table_id;
+  x_google_request_params_ = std::make_pair("x-goog-request-params", value);
+}
+
+MetadataUpdatePolicy::MetadataUpdatePolicy(MetadataUpdatePolicy const& rhs) {
+  google_cloud_resource_prefix_ = rhs.google_cloud_resource_prefix_;
+  x_google_request_params_ = rhs.x_google_request_params_;
+}
+
+void MetadataUpdatePolicy::setup(grpc::ClientContext& context) const {
+  context.AddMetadata(google_cloud_resource_prefix_.first,
```

Comment:
Sorry for the last minute change, but we should delete `google_cloud_resource_prefix_`.  Nice job encapsulating the code, it will be a relatively small change.

Reply:
changing it, but why do we want to remove it?

Reply:
See the comments on #190.  The `google-cloud-resource-prefix` was an ad-hoc thing for Bigtable, and was in use when we wrote the original bug.  `x-goog-request-params` is the new thing that will be used across multiple (I think all) services.  Igor just finished changing Cloud Bigtable to accept the new thing, so we no longer need the old, and we can save a few bytes on the wire if we send only one.
Perfectly good question to ask btw.